### PR TITLE
[PS-93] Distributed Ip rate limiting

### PR DESCRIPTION
**Pull request by shane-melton**
## Type of change

- [X] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other

## Objective
Update Api and Identity projects to use a distributed storage medium (Redis) for tracking Ip rate limiting counts. Currently, they use an in memory storage, which does not properly block requests when deployed at scale (multiple instances store their own individual access counts, w/o knowledge of others).

Before being merged/deployed a new Redis connection string needs to be added to `GlobalSettings` under the key `globalSettings:redis:connectionString` that points towards an available Redis instance, otherwise the API/Identity projects will fall back to an in memory store/cache. 

### Package Changes for src/Core/Core.csproj:
* Upgrade `AspNetCoreRateLimit` to 4.0.2 and add `AspNetCoreRateLimit.Redis@1.0.1` to gain Redis support. 
* Replace/upgrade `Microsoft.Extensions.Caching.Redis` with `Microsoft.Extensions.Caching.StackExchangeRedis` as the former is no longer supported and has dependency conflicts with `AspNetCoreRateLimit@4.0.2`
* Upgrade `NewtonSoft.Json` to 13.0.1 as it is a required for the latest `AspNetCoreRateLimit` package. I did not find any breaking changes with the upgrade, and I know we are moving away from NewtonSoft.Json internally so I don't believe it will be a problem.

### Other Changes:
* **src/[Api | Identity]/Startup.cs:** Configure necessary services for Ip rate limiting and the distributed Redis cache

* **src/Core/HostedServices/IpRateLimitSeedStartupService.cs:** Added a hosted service to seed any Ip rate limit policies on initial application startup. (Avoids modifying `Program.cs`, which can get messy)

* **src/SharedWeb/Utilities/ServiceCollectionExtensions.cs:** Add two extensions to consistently add distributed caching and ip rate limiting across all projects. Remove the distributed cache service configuration out of `AddDistributedIdentityServices()` as those are now configured by `AddDistributedCache()`

* **src/Core/Settings/GlobalSettings.cs:** Add a _general_ Redis connection string that is used by the new service collection extensions to add distributed caching to the project. Having a single Redis instance shared between all the projects should make the infrastructure requirements and configurations simpler/cleaner going forward. 

* **src/Core/Utilities/CustomIpRateLimitMiddleware.cs:** Updated for breaking changes due to `AspNetCoreRateLimit` package upgrade. Now using an `IDistributedCache` for storing/retrieving IP block counts.

* **src/Core/Utilities/DistributedCacheExtensions.cs:** Some helper extensions when interacting with the `IDistributedCache` service.

## Before you submit

- [x] I have checked for formatting errors (`dotnet tool run dotnet-format --check`) (required)
- [ ] If making database changes - I have also updated Entity Framework queries and/or migrations
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [X] This change has particular **deployment requirements** (notify the DevOps team)

### CODE REVIEW

File: src/Core/Utilities/CustomIpRateLimitMiddleware.cs [hunk 1]
```diff
@@ -63,28 +64,19 @@ public override Task ReturnQuotaExceededResponse(HttpContext httpContext, RateLi
...
-            {
-                s += $"Header \"{header.Key}\": {header.Value} \n";
-            }
+            foreach (var header in httpContext.Request.Headers) s += $"Header \"{header.Key}\": {header.Value} \n";
 
-            foreach (var query in httpContext.Request.Query)
-            {
-                s += $"Query \"{query.Key}\": {query.Value} \n";
-            }
+            foreach (var query in httpContext.Request.Query) s += $"Query \"{query.Key}\": {query.Value} \n";
```
Comment (justindbaur) [inline src/Core/Utilities/CustomIpRateLimitMiddleware.cs:79]: nit: We use curly braces even on single line statements. https://contributing.bitwarden.com/code-style/#control-blocks

File: src/Core/Utilities/CustomIpRateLimitMiddleware.cs [hunk 2]
```diff
@@ -4,53 +4,54 @@
...
             _logger = logger;
         }
 
         public override Task ReturnQuotaExceededResponse(HttpContext httpContext, RateLimitRule rule, string retryAfter)
         {
-            var message = string.IsNullOrWhiteSpace(_options.QuotaExceededMessage) ?
-                $"Slow down! Too many requests. Try again in {rule.Period}." : _options.QuotaExceededMessage;
+            var message = string.IsNullOrWhiteSpace(_options.QuotaExceededMessage)
+                ? $"Slow down! Too many requests. Try again in {rule.Period}."
+                : _options.QuotaExceededMessage;
```
Comment (justindbaur) [inline src/Core/Utilities/CustomIpRateLimitMiddleware.cs:41]: I like this refactor.

File: src/Core/Utilities/DistributedCacheExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        }
+
+        public static Task SetAsync<T>(this IDistributedCache cache, string key, T value,
+            DistributedCacheEntryOptions options)
+        {
+            var bytes = Encoding.UTF8.GetBytes(JsonHelpers.LegacySerialize(value));
+            return cache.SetAsync(key, bytes, options);
+        }
+
+        public static bool TryGetValue<T>(this IDistributedCache cache, string key, out T value)
```
Comment (justindbaur) [inline src/Core/Utilities/DistributedCacheExtensions.cs:34]: I love the `bool Try... out` pattern ❤️

**Review by justindbaur — CHANGES_REQUESTED** [review]

**Review by justindbaur — COMMENTED** [review]
Looks good, looks like some tests are failing though. Might need to rebase off master. 

@mimartin12 Are we able to test this in QA?

**Comment by shane-melton** [comment]
That failing test is fixed now

**Review by mimartin12 — APPROVED** [review]

**Pull request merged**